ui/src/components/stream-control.tsx

```diff
@@ -1,95 +1,100 @@
+          let localStream = new MediaStream();
+          let clonedIds = new Set();
+          let lastParentStream = null;
+          function setStatus(msg){ if(statusEl) statusEl.textContent = msg; }
+          function validateOpener(){
+            try {
+              if(!window.opener || window.opener !== openerRef){ setStatus('Opener lost. Closing…'); setTimeout(()=>window.close(),800); return false; }
+              void window.opener.location.href; // access for same-origin check
+              return true;
+            } catch { setStatus('Cross-origin opener. Closing…'); setTimeout(()=>window.close(),800); return false; }
+          }
+          function attachVideo(){ if(video.srcObject !== localStream) video.srcObject = localStream; }
+          function cloneTracks(){
+            if(!validateOpener()) return;
+            const parentStream = window.opener.__comfystreamRemoteStream;
+            if(!parentStream){ setStatus('Waiting for stream…'); return; }
+            if(lastParentStream && lastParentStream !== parentStream){
+              localStream.getTracks().forEach(t=>{ try{t.stop();}catch{} });
+              localStream = new MediaStream();
+              clonedIds = new Set();
+            }
+            lastParentStream = parentStream;
+            let added = false;
+            parentStream.getTracks().forEach(src => {
+              if(src.readyState === 'ended') return;
+              if(!clonedIds.has(src.id)){
+                try{ const c = src.clone(); localStream.addTrack(c); clonedIds.add(src.id); added = true; c.addEventListener('ended',()=>{ clonedIds.delete(src.id); }); }
+                catch{ /* fallback skip */ }
+              }
+            });
+            if(added){ attachVideo(); setStatus('Live'); if(video.play) video.play().catch(()=>{}); }
+          }
+          const interval = setInterval(()=>{
+            attempts++;
+            if(!validateOpener()){ clearInterval(interval); return; }
+            // Parent cleared global (disconnect)
+            if(!window.opener.__comfystreamRemoteStream){ setStatus('Parent stream ended'); clearInterval(interval); setTimeout(()=>window.close(),1200); return; }
+            cloneTracks();
+            // Remove ended local tracks (allow re-clone)
+            localStream.getTracks().forEach(t=>{ if(t.readyState==='ended'){ localStream.removeTrack(t); try{t.stop();}catch{}; clonedIds.delete(t.id); }});
+            if(attempts>=MAX_ATTEMPTS && localStream.getTracks().length===0){ setStatus('Timeout waiting for stream'); clearInterval(interval); setTimeout(()=>window.close(),1500); }
+          },300);
+          window.addEventListener('beforeunload', ()=>{ clearInterval(interval); localStream.getTracks().forEach(t=>{ try{t.stop();}catch{} }); });
+          // Initial attempt
+          cloneTracks();
+        })();
+      </script>
+    </body></html>`;
+    popup.document.write(html);
+    popup.document.close();
+  }, []);
```

Comment:
Instead of using `document.write()`, can we add a proper nodejs page for the pop-out window? Then it can compile to `stream.html` with `npm run build`